docs: benchmarking (#270)
jina-ai/docarray: docs/advanced/document-store/benchmarking.md

> # Benchmarking

Reviewer: ```suggestion
# Benchmark
```

> ## Results

Reviewer: ```suggestion
```

> The script `scripts/benchmarking.py` benchmarks DocArray's supported Document Stores in 6 different operations:

Reviewer: User first-person.

"To understand ..., we benchmark ... in ... way. In particular, we care about ... , which lists as follows:"

Reviewer: ```
````{tip}
To produce the result, one can run ...
````

```

> * Create (indexing Documents)

Reviewer: ```suggestion
- **Create**: remove () and really explain what does that mean by create, what function do you call. Right now you are explain `create` by `indexing Documents`, but then what is indexing document? 
```

Same for all other ops, explain what exactly happens there.

Author: explanation added

* Read
* Update
* Delete
* Find Document by vector (Nearest Neighbor Search or Approximate Nearest Neighbor Search depending on the `backend`)
* Find Document by condition (apply filter)

> ````{tab} #1: Same HNSW parameters

Reviewer: ```suggestion
````{tab} Same HNSW parameters
```

| Backend       | Create (s) | Read (ms) | Update (ms) | Delete (ms) | Find by vector (s) | Recall at k=10 for | Find by condition (s) |
|---------------|------------|-----------|-------------|-------------|--------------------|--------------------|-----------------------|

> | Memory        | 0.69       | 0.13      | 0.014       | 0.12        | 1.6                | 1                  | 5.4                   |

Reviewer: - Number column should be right-align
- Fixed-length, for example, 0.014 vs 44 is not professional. In this case you may want to change unit`ms` to `ns`

> | Memory        | 0.69       | 0.13      | 0.014       | 0.12        | 1.6                | 1                  | 5.4                   |

Reviewer: `5.4` but all others are integers?

| ElasticSearch | 720        | 8.98      | 41          | 65          | 0.29               | 0.87               | 704                   |

````

| Backend       | Create (s) | Read (ms) | Update (ms) | Delete (ms) | Find by vector (s) | Recall at k=10 for | Find by condition (s) |
|---------------|------------|-----------|-------------|-------------|--------------------|--------------------|-----------------------|
| Memory        | 0.63       | 0.12      | 0.013       | 0.15        | 1.42               | 1                  | 5                     |
| Qdrant        | 2192       | 2         | 1.7         | 10          | 0.008              | 0.51               | 608                   |
| Sqlite        | 4422       | 0.32      | 0.36        | 1           | 24                 | 1                  | 31                    |
| Annlite       | 111        | 0.24      | 8           | 31          | 0.41               | 0.14               | 30                    |
| Weaviate      | 1609       | 2.14      | 32.28       | 26          | 0.005              | 0.1                | 1176                  |

> | ElasticSearch | 719        | 9.52      | 29.77       | 50          | 0.23               | 0.83               | 1183                  |

Reviewer: what is this? a tab?

Author: fixed

> ## Experiment Settings

Reviewer: ```suggestion
## Experimental setup
```

All non-h1 follow Capital-case, not CamelCase

Since most of these Document Stores use their implementation of the HNSW Approximate Nearest Neighbor Search algorithm, 
with various default HNSW parameters, we conducted 2 benchmarking experiments for the `Find By Vector` operation:
1. Set up the Document Stores with the same HNSW parameters 
2. Set up the Document Stores with their default HNSW parameters at the time the benchmarking experiment was conducted.

> Furthermore, we provide the `Recall At K` value, considering the exhaustive search as the ground truth. This allows 

Reviewer: ```suggestion
Furthermore, we provide the `Recall@K` value, considering the exhaustive search as the ground truth. This allows 
```

you to also take into consideration the quality, not just the speed.

```{important}
The Sqlite and the in-memory Document Stores ** do not implement ** approximate nearest neighbor search and offer 
exhaustive search instead. That's why, they give the maximum quality but are the slowest when it comes to search nearest 
neighbors.
```

The results were conducted on a 4.5 Ghz AMD Ryzen Threadripper 3960X 24-Core Processor with Python 3.8.5 and using the official docker 
images of the storage backends. The docker images were allocated 40 GB of RAM.

The benchmarking experiments used the following parameters:
* Number of indexed Documents: 1M
* Number of query Documents: 1
* Embedding dimension: 128
* Number of Documents (K) to be retrieved for `Find Document by vector`: 10

For the first experiment, we fixed the following HNSW parameters for the Document Stores that support ANN:

> * `ef` (construction): 100
> * `ef` (search): 100
> * `m` (max connections): 16

Reviewer: ```suggestion
* `ef=100` at construction ...
* `ef=100` Please elaborate it
* `m` (max connections): 16, same format and [lease elaborate it
```

Author: explained

some offer better quality. Therefore, we try to benchmark on 6 operations (CRUD + Find by vector + Find by condition)
and report quality measurement (Recall at K).
* Not just the speed, quality in mind: Since some backends offer better performance and some offer better quality, 
we make sure to report the quality measurement for the Approximate Nearest Neighbor Search. This will allow users to 
choose the backend that best suit their cases.
* Same experiment, same API: DocArray offers the same API across all backends and therefore, we built on top of it the 
same benchmarking experiment. Furthermore, we made sure to fix the same HNSW parameters for backends that support 
Approximate Nearest Neighbor Search. All backends are run on official Docker containers, local to the DocArray client 
which allows having similar network overhead. We also allocate the same resources for those Docker containers and all 
servers are run in a single node setup.
* Benefit the user as much as possible: We offer the same conditions and resources to all backends, but our experiment 
favours the backends that use the resources efficiently. Therefore, some backends might not use the network, or use 

> GRPC instead of HTTP, use batch operations,... And we're okay with that, as long as it benefits the DocArray and Jina 

Reviewer: ```suggestion
GRPC instead of HTTP, or use batch operations. And we're okay with that, as long as it benefits the DocArray and Jina 
```

user.
* Open to improvements: We are constantly improving the performance of storage backends from the DocArray side and 
updating the benchmarks accordingly. If you believe we missed an optimization (perform an operation in batches, benefit 
from a recent feature in upstream, avoid unnecessary steps,...), feel free to raise a PR or issue, we're open to 
your contributions!

## Backends covered
Our benchmarks cover all backends supported by DocArray:
* In-memory
* Sqlite
* Qdrant
* Annlite
* Weaviate

> * ElasticSearch

Reviewer: Add in-doc link to each, link to each's corresponding page in our docs.

Author: done

We do not cover the following backends for those reasons:
* Milvus: currently DocArray does not integrate with Milvus. We're open for contributions to DocArray's repository to 
support it.
* Pinecone: Pinecode support is coming soon, and we'll add it to these benchmarks once available.
* Faiss, Annoy, Scann: We do not benchmark algorithms or ANN libraries. We only benchmark backends that can be used as 
Document stores. Actually we do not benchmark HNSW itself, but it is used by some backends internally.

## Conclusion

> To conclude, you can look at those benchmarks and select the backend that suits better your case.

Reviewer: This is too weak.

"To concolude, if you are working under ?? Documents, use ??; if you are ...., use ..."

> If you're playing around on a small dataset, you can use `memory` as storage.

Reviewer: how small is small, 10, 100? or 1? this sentence is useless without the number

> If you don't care much about the Nearest Neighbor Search speed and you're experiment with larger data, you can use`

Reviewer: ```suggestion
If you don't care much about the  speed of xxxx and you're experiment with larger data, you can use`
```
